File: app/src/main/java/de/bitmacht/workingtitle36/Value.java
package de.bitmacht.workingtitle36;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.util.Pair;

import java.text.DecimalFormatSymbols;
import java.util.Collection;
import java.util.Currency;
import java.util.Locale;

/**
 * Combines an amount and a ISO 4217 currency code
 */
public class Value implements Parcelable {
    /**
     * The ISO 4217 currency code associated with this object
     */
    public final String currencyCode;
    /**
     * The number of minor currency units represented by this object
     */
    public final long amount;

    /**
     * Create a new instance of Value
     * @param amount The number of minor currency units
     * @param currencyCode The ISO 4217 currency code
     */
    public Value(@NonNull String currencyCode, long amount) {
        this.currencyCode = currencyCode;
        this.amount = amount;
    }

    /**
     * Returns a new Value having the currency of this amount, but with a different amount
     */
    public Value withAmount(long amount) {
        return new Value(this.currencyCode, amount);
    }

    /**
     * Returns the Currency associated with this object's currency code
     */
    public Currency getCurrency() {
        return Currency.getInstance(currencyCode);
    }

    /**
     * Returns a new Value with the sum of this and the other Value
     * @throws CurrencyMismatchException If the currencies differ
     */
    public Value add(Value other) throws CurrencyMismatchException {
        if (!currencyCode.equals(other.currencyCode)) {
            throw new CurrencyMismatchException("Unable to add " + other.currencyCode + " to " + currencyCode);
        }
        return new Value(currencyCode, amount + other.amount);
    }

    /**
     * Return a new Value with the sum of this and all Values from others
     * @throws CurrencyMismatchException If any currency code differs from this
     */
    public Value addAll(Value[] others) throws CurrencyMismatchException {
        long sumount = amount;
        for (Value other : others) {
            if (!currencyCode.equals(other.currencyCode)) {
                throw new CurrencyMismatchException("Unable to add " + other.currencyCode + " to " + currencyCode);
            }
            sumount += other.amount;
        }
        return new Value(currencyCode, sumount);
    }

    /**
     * Return a new Value with the sum of this and all Values from others
     * @throws CurrencyMismatchException If any currency code differs from this
     */
    public Value addAll(Collection<Value> others) throws CurrencyMismatchException {
        long sumount = amount;
        for (Value other : others) {
            if (!currencyCode.equals(other.currencyCode)) {
                throw new CurrencyMismatchException("Unable to add " + other.currencyCode + " to " + currencyCode);
            }
            sumount += other.amount;
        }
        return new Value(currencyCode, sumount);
    }

    /**
     * Builds an appropriate String for a given currency and amount.
     * Uses the default locale.
     * @return The generated String
     */
    public String getString() {
        return getString(Locale.getDefault());
    }

    /**
     * Builds an appropriate String for a given currency and amount.
     * @param locale The Locale to use for the decimal separator
     * @return The generated String
     */
    public String getString(Locale locale) {
        Pair<String, String> vs = getValueAndSymbolStrings(locale);
        return vs.first + vs.second;
    }

    public Pair<String, String> getValueAndSymbolStrings(Locale locale) {
        Currency currency = getCurrency();
        int fractionDigits = currency.getDefaultFractionDigits();
        String symbol = currency.getSymbol();
        char decimalSeparator = DecimalFormatSymbols.getInstance(locale).getMonetaryDecimalSeparator();

        String sign = amount < 0 ? "-" : "";
        long umount = Math.abs(amount);
        long major = umount;
        long minor = 0;
        boolean showFraction = fractionDigits > 0;
        if (showFraction) {
            long div = (long) Math.pow(10, fractionDigits);
            major = umount / div;
            minor = umount % div;
        }

        String valueString = showFraction ? String.format("%s%d%s%02d", sign, major, decimalSeparator, minor) :
                String.format("%s%d", sign, major);

        return new Pair<>(valueString, symbol);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(currencyCode);
        dest.writeLong(amount);
    }

    public static final Parcelable.Creator<Value> CREATOR =
            new Parcelable.Creator<Value>() {
                @Override
                public Value createFromParcel(Parcel source) {
                    return new Value(source);
                }

                @Override
                public Value[] newArray(int size) {
                    return new Value[size];
                }
            };

    private Value(Parcel in) {
        currencyCode = in.readString();
        amount = in.readLong();
    }

    public static class CurrencyMismatchException extends Exception {
        public CurrencyMismatchException(String s) {
            super(s);
        }
    }

    @Override
    public String toString() {
        return currencyCode + ":" + amount;
    }
}

File: app/src/main/java/de/bitmacht/workingtitle36/Value.kt
/*
 * Copyright 2016 Kamil Sartys
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package de.bitmacht.workingtitle36

import android.os.Parcel
import android.os.Parcelable
import android.util.Pair
import java.text.DecimalFormatSymbols
import java.util.*

/**
 * Combines an amount and a ISO 4217 currency code
 *
 * @param currencyCode The ISO 4217 currency code
 * @param amount The number of minor currency units
 */
class Value(val currencyCode: String, val amount: Long = 0) : Parcelable {

    /**
     * Returns a new Value having the currency of this amount, but with a different amount
     */
    fun withAmount(amount: Long): Value = Value(this.currencyCode, amount)

    /**
     * Returns the Currency associated with this object's currency code
     */
    val currency: Currency by lazy { Currency.getInstance(currencyCode) }

    /**
     * Returns a new Value with the sum of this and the other Value
     * @throws CurrencyMismatchException If the currencies differ
     */
    @Throws(CurrencyMismatchException::class)
    fun add(other: Value): Value {
        if (!isSameCurrency(other))
            throw CurrencyMismatchException("Unable to add ${other.currencyCode} to $currencyCode")
        return withAmount(amount + other.amount)
    }

    /**
     * Return a new Value with the sum of this and all Values from others having the same currency
     */
    fun addAll(others: Collection<Value>): Value {
        return withAmount(others.filter { other -> isSameCurrency(other) }
                .fold(amount, { acc, other -> acc + other.amount }))
    }

    /**
     * Builds an appropriate String for a given currency and amount.
     * Uses the default locale.
     * @return The generated String
     */
    val string: String
        get() = getString(Locale.getDefault())

    /**
     * Builds an appropriate String for a given currency and amount.
     * @param locale The Locale to use for the decimal separator
     *
     * @return The generated String
     */
    fun getString(locale: Locale): String {
        val vs = getValueAndSymbolStrings(locale)
        return vs.first + vs.second
    }

    fun getValueAndSymbolStrings(locale: Locale): Pair<String, String> {
        val fractionDigits = currency.defaultFractionDigits
        val symbol = currency.symbol
        val decimalSeparator = DecimalFormatSymbols.getInstance(locale).monetaryDecimalSeparator

        val sign = if (amount < 0) "-" else ""
        val umount = Math.abs(amount)
        var major = umount
        var minor: Long = 0
        val showFraction = fractionDigits > 0
        if (showFraction) {
            val div = Math.pow(10.0, fractionDigits.toDouble()).toLong()
            major = umount / div
            minor = umount % div
        }

        val valueString = if (showFraction)
            String.format("%s%d%s%02d", sign, major, decimalSeparator, minor)
        else
            String.format("%s%d", sign, major)

        return Pair(valueString, symbol)
    }

    fun isSameCurrency(other: Value): Boolean {
        return other.currencyCode == currencyCode
    }

    override fun describeContents(): Int = 0

    override fun writeToParcel(dest: Parcel, flags: Int) = with(dest) {
        writeString(currencyCode)
        writeLong(amount)
    }

    private constructor(src: Parcel) : this(src.readString(), src.readLong())

    class CurrencyMismatchException(s: String) : Exception(s)

    override fun toString(): String = "$currencyCode : $amount"

    companion object {

        @JvmField val CREATOR: Parcelable.Creator<Value> = object : Parcelable.Creator<Value> {
            override fun createFromParcel(source: Parcel): Value = Value(source)
            override fun newArray(size: Int): Array<Value?> = arrayOfNulls(size)
        }
    }
}
